src/kernel/pagecache.c

```diff
@@ -533,41 +546,56 @@ closure_function(5, 1, void, pagecache_write_sg_finish,
         goto exit;
 
     pagecache_lock_node(pn);
-    pagecache_page pp = page_lookup_nodelocked(pn, pi);
+    pagecache_page pp = page_lookup_nodelocked(pn, bound(pi));
 
     /* copy data to the page cache */
-    u64 offset = q.start & MASK(page_order);
+    u64 offset = (bound(pi) == (q.start >> page_order)) ? (q.start & MASK(page_order)) : 0;
     range r = irange(q.start & ~MASK(block_order), q.end);
 #ifdef KERNEL
     context saved_ctx = bound(saved_ctx);
     if (saved_ctx)
         use_fault_handler(saved_ctx->fault_handler);
 #endif
     do {
-        assert(pp != INVALID_ADDRESS && page_offset(pp) == pi);
-        u64 copy_len = MIN(q.end - (pi << page_order), cache_pagesize(pc)) - offset;
+        assert(pp != INVALID_ADDRESS && page_offset(pp) == bound(pi));
+        u64 copy_len = MIN(q.end - (bound(pi) << page_order), cache_pagesize(pc)) - offset;
+        if (sg)
+            sg_fault_in(sg, copy_len);  /* to prevent page faults while the state lock is held */
+        pagecache_lock_state(pc);
+        if (page_state(pp) == PAGECACHE_PAGESTATE_READING) {
+            /* A read request occurred in the middle of this write: postpone the completion of this
+             * write so that the data being written will overwrite the data fetched by the read
+             * request. */
+            enqueue_page_completion_statelocked(pc, pp, (status_handler)closure_self());
+            pagecache_unlock_state(pc);
+            break;
+        }
         if (sg) {
             u64 res = sg_copy_to_buf(pp->kvirt + offset, sg, copy_len);
             assert(res == copy_len);
         } else {
             zero(pp->kvirt + offset, copy_len);
         }
-        pagecache_lock_state(pc);
-        assert(page_state(pp) != PAGECACHE_PAGESTATE_READING);
         if (page_state(pp) != PAGECACHE_PAGESTATE_DIRTY)
+            /* Don't release the page, otherwise if it has been marked for eviction it will be
+             * evicted while still dirty. */
             change_page_state_locked(pc, pp, PAGECACHE_PAGESTATE_DIRTY);
+        else
+            pagecache_page_release_locked(pc, pp);
```

Comment:
There is still a refcount problem relating to write_sg I think. If I merge this branch into the low memory branch, I can cause an assertion failure running the fs_full test with lower than normal memory (e.g. 384M) where the page state of a page when going to refcount 0 is PAGECACHE_PAGESTATE_NEW instead of FREE. I can make it pass by unconditionally calling pagecache_page_release_locked and not changing the state to DIRTY, but that's not correct for the page state.

Reply:
I could reproduce the issue after merging the low memory branch into this branch. Assuming you resolved the merge conflicts in the same way as I did, the issue is caused by the code in pagecache_write_sg() that changes the page state to WRITING (at line 696 in your branch); removing that code fixes the issue.

Reply:
You're right -- I forgot to remove those lines after the merge, and it does also make it work for me. Looks good, thanks!